frontend/lib/types/Param.cpp

```diff
@@ -89,157 +90,168 @@ bool Param::isParamOpFoldable(chpl::uast::PrimitiveTag op) {
   }
 }
 
+
+/*
+These helpers get the immediate value from a Param in the proper type if the
+param exists. Otherwise, they return an empty value of the proper type.
+*/
+template<typename T, typename S>
+static T getImmediateValueOrEmpty(const S* p) {
+  if (p) {
+    return p->value();
+  }
+  // handle complex and string types in overloads below
+  return (T) 0;
+}
+
+static Param::ComplexDouble getImmediateValueOrEmpty(const ComplexParam* p) {
+  if (p) {
+    return p->value();
+  }
+  return Param::ComplexDouble(0.0, 0.0);
+}
+
+static ImmString getImmediateValueOrEmpty(const StringParam* p) {
+  if (p) {
+    return p->value().podUniqueString();
+  }
+  return UniqueString().podUniqueString();
+}
+
+/*
+  Get the immediate value from a Param in the proper type. Pass nullptr as param
+  to get an empty immediate for a param of Type t.
+*/
 static
 Immediate paramToImmediate(const Param* p, const Type* t) {
   Immediate ret;
 
-  switch (p->tag()) {
-    case paramtags::BoolParam:
-      {
-        auto bp = (const BoolParam*) p;
-        auto v = bp->value();
-        ret.const_kind = NUM_KIND_BOOL;
-        ret.num_index = BOOL_SIZE_SYS;
-        ret.v_bool = v;
-        return ret;
-      }
-    case paramtags::ComplexParam:
-      {
-        auto cp = (const ComplexParam*) p;
-        auto v = cp->value();
-        auto ct = t->toComplexType();
-        CHPL_ASSERT(ct);
-        if (ct == nullptr) return ret;
-        ret.const_kind = NUM_KIND_COMPLEX;
-        if (ct->bitwidth() == 64) {
-          ret.num_index = COMPLEX_SIZE_64;
-          ret.v_complex64.r = v.re;
-          ret.v_complex64.i = v.im;
-        } else if (ct->bitwidth() == 128) {
-          ret.num_index = COMPLEX_SIZE_128;
-          ret.v_complex128.r = v.re;
-          ret.v_complex128.i = v.im;
-        } else {
-          CHPL_ASSERT(false && "case not handled");
-        }
-        return ret;
-      }
-    case paramtags::EnumParam:
-      {
+  if (t->isBoolType()) {
```

Comment:
I don't think we should assume here that if `t` is a bool type that `p` is a bool param. So it seems to me that the switch was better (at least, the blocks here should perhaps be based on the param type first? I would believe you might have run into a case where the type `t` needs to be used that wasn't working yet.